substrate/frame/support/src/generate_genesis_config.rs

```diff
@@ -0,0 +1,875 @@
+/// struct RuntimeGenesisConfig {
+///     a_field: u32,
+///     b_field: B,
+///     c_field: u32,
+/// }
+///
+/// #[derive(Default, serde::Serialize, serde::Deserialize)]
+/// #[serde(rename_all = "camelCase")]
+/// struct B {
+/// 	i_field: u32,
+/// 	j_field: u32,
+/// }
+///
+/// impl B {
+/// 	fn new() -> Self {
+/// 		Self { i_field: 0, j_field: 2 }
+/// 	}
+/// }
+///
+/// assert_eq!(
+/// 	build_struct_json_patch! ( RuntimeGenesisConfig {
+/// 		a_field: 66,
+/// 	}),
+/// 	serde_json::json!({
+/// 			"aField": 66,
+/// 	})
+/// );
+///
+/// assert_eq!(
+/// 	build_struct_json_patch! ( RuntimeGenesisConfig {
+/// 		//"partial" initialization of `b_field`
+/// 		b_field: B {
+/// 			i_field: 2,
+/// 		}
+/// 	}),
+///
+/// 	serde_json::json!({
+/// 		"bField": {"iField": 2}
+/// 	})
+/// );
+///
+/// assert_eq!(
+/// 	build_struct_json_patch! ( RuntimeGenesisConfig {
+/// 		a_field: 66,
+/// 		//"full" initialization of `b_field`
+/// 		b_field: B::new()
+/// 	}),
+/// 	serde_json::json!({
+/// 		"aField": 66,
+/// 		"bField": {"iField": 0, "jField": 2}
+/// 	})
+/// );
+/// ```
+///
+/// In this example:
+/// ```ignore
+/// 	build_struct_json_patch! ( RuntimeGenesisConfig {
+/// 		b_field: B {
+/// 			i_field: 2,
+/// 		}
+/// 	}),
+/// ```
+/// `b_field` is partially initialized, it will be expanded to:
+/// ```ignore
+/// RuntimeGenesisConfig {
+/// 		b_field {
+/// 			i_field: 2,
+/// 			..Default::default()
+/// 		},
+/// 		..Default::default()
+/// }
+/// ```
+/// while all other fields are initialized with default values. The macro serializes this, retaining
+/// only the provided fields.
+#[macro_export]
+macro_rules! build_struct_json_patch {
+	(
+		$struct_type:ident { $($tail:tt)* }
+	) => {
+		{
+			use $crate::generate_genesis_config::{InitializedField, retain_initialized_fields};
+			extern crate alloc;
+			use alloc::{string::ToString, vec::Vec };
+			let mut keys : Vec<InitializedField> = vec![];
+			#[allow(clippy::needless_update)]
+			let struct_instance = build_struct_json_patch!($struct_type, keys @  { $($tail)* });
+			let mut json_value =
+				serde_json::to_value(struct_instance).expect("serialization to json should work. qed");
+			retain_initialized_fields(&mut json_value, &keys, Default::default());
+			json_value
+		}
+	};
+	($struct_type:ident, $all_keys:ident @ { $($tail:tt)* }) => {
+		$struct_type {
+			..build_struct_json_patch!($struct_type, $all_keys @ $($tail)*)
+		}
+	};
+	($struct_type:ident, $all_keys:ident  @  $key:ident: $type:tt { $keyi:ident : $value:tt }  ) => {
+		$struct_type {
+			$key: {
+				$all_keys.push(InitializedField::Partial(stringify!($key).to_string()));
+				$all_keys.push(
+					InitializedField::Full(alloc::format!("{}.{}", stringify!($key), stringify!($keyi)))
+				);
+				$type {
+					$keyi:$value,
+					..Default::default()
+				}
+			},
+			..Default::default()
+		}
+	};
+	($struct_type:ident, $all_keys:ident  @  $key:ident:  $type:tt { $($body:tt)* } ) => {
```

Comment:
```suggestion
	($struct_type:ident, $all_keys:ident  @  $key:ident:  $type:ident { $($body:tt)* } ) => {
```
I think type can only be a type identifier here.

Reply:
yeah, right it can be identifier.

Reply:
done in: 0c24c455